eventdemo/source/c/src/kaa_demo.c

```diff
@@ -19,212 +19,155 @@
-    response->thermostat_info->data = info;
-
-    int32_t *current_degree = (int32_t *) KAA_MALLOC(sizeof(int32_t));
-    *current_degree = -5;
+    room_t *head;
+    room_t *tail;
+} list_t;
 
-    int32_t *target_degree = (int32_t *) KAA_MALLOC(sizeof(int32_t));
-    *target_degree = 10;
+list_t lst;
 
-    info->degree = kaa_thermostat_event_class_family_union_int_or_null_branch_0_create();
-    info->degree->data = current_degree;
-
-    info->target_degree = kaa_thermostat_event_class_family_union_int_or_null_branch_0_create();
-    info->target_degree->data = target_degree;
-
-    info->is_set_manually = kaa_thermostat_event_class_family_union_boolean_or_null_branch_1_create();
-
-    kaa_event_manager_send_kaa_thermostat_event_class_family_thermostat_info_response(
-                                kaa_client_get_context(kaa_client)->event_manager, response, NULL);
-
-    response->destroy(response); // Destroying event that was successfully sent
+static kaa_client_t *kaa_client = NULL;
 
-    event->destroy(event);
+void initialize_List(list_t *lst)
+{
+    lst->head = NULL;
+    lst->tail = NULL;
 }
 
-
-void kaa_on_thermostat_info_response(void *context
-                                   , kaa_thermostat_event_class_family_thermostat_info_response_t *event
-                                   , kaa_endpoint_id_p source)
+void push(list_t *lst, const char *room_name)
 {
-    KAA_DEMO_UNUSED(context);
-    KAA_DEMO_UNUSED(source);
-
-    printf("ThermostatInfoResponse event received!\n");
-
-    if (event->thermostat_info->type == KAA_THERMOSTAT_EVENT_CLASS_FAMILY_UNION_THERMOSTAT_INFO_OR_NULL_BRANCH_0) {
-        kaa_thermostat_event_class_family_thermostat_info_t *info =
-                (kaa_thermostat_event_class_family_thermostat_info_t *) event->thermostat_info->data;
-
-        if (info->degree->type == KAA_THERMOSTAT_EVENT_CLASS_FAMILY_UNION_INT_OR_NULL_BRANCH_0) {
-            int32_t *degree = (int32_t *) info->degree->data;
-            printf("Degree=%d\n", *degree);
-        }
-        if (info->target_degree->type == KAA_THERMOSTAT_EVENT_CLASS_FAMILY_UNION_INT_OR_NULL_BRANCH_0) {
-            int32_t *target_degree = (int32_t *) info->target_degree->data;
-            printf("Target degree=%d\n", *target_degree);
-        }
-        if (info->is_set_manually->type == KAA_THERMOSTAT_EVENT_CLASS_FAMILY_UNION_BOOLEAN_OR_NULL_BRANCH_0) {
-            int8_t *is_set_manually = (int8_t *) info->is_set_manually->data;
-            printf("Is_set_manually=%s\n", (*is_set_manually) ? "true" : "false");
-        }
+    pthread_mutex_lock(&lock);
+    room_t *t = malloc(sizeof(room_t));
+    if (t == NULL) {
+        printf("Out of memory\n");
+        exit(EXIT_FAILURE);
     }
-
-    event->destroy(event);
+    strncpy(t->room_name, room_name, strlen(room_name) + 1);
+    t->next = NULL;
+    if (lst->head == NULL)
```

Comment:
Please re-write this block to 
if (...) {
...
} else {
...
} 
mutex_unlock();

This will eliminate dual-point return and mutex unlocking as well as readability.